src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,682 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A type interface for a range of bits.
+#[allow(clippy::exhaustive_structs)]
+#[derive(Debug, Clone)]
+pub struct BitRange<'a, T, const START: u8, const END: u8>(pub &'a T);
+
+/// A type interface for a range of bits.
+#[allow(clippy::exhaustive_structs)]
+#[derive(Debug)]
+pub struct BitRangeMut<'a, T, const START: u8, const END: u8>(pub &'a mut T);
+
+// std::fmt::Display
+macro_rules! bit_range_display {
+    ($x:ty) => {
+        impl<const START: u8, const END: u8> std::fmt::Display for BitRange<'_, $x, START, END> {
+            #[doc = concat!("
+                ```
+                use bit_fields::BitRange;
+                let x = 18", stringify!($x), ";
+                assert_eq!(BitRange::<_,0,4>(&x).to_string(),2", stringify!($x),".to_string());
+                assert_eq!(BitRange::<_,4,8>(&x).to_string(),1", stringify!($x),".to_string());
+                ```
+            ")]
+            #[inline]
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "{}", <$x>::from(self))
+            }
+        }
+    };
+}
+
+bit_range_display!(u128);
+bit_range_display!(u64);
+bit_range_display!(u32);
+bit_range_display!(u16);
+bit_range_display!(u8);
```

Comment:
The macro generates test code contributing to documentation and coverage.

Although typically generics should be preferable to macros, using generics in this case produces more complex code, less well documented code and worse code coverage. These cases are fairly common, you may find them in the standard library where generics could be used but are unfortunately non-ideal.

I  did consider using generics for some of these but ultimately I think it negatively affects code quality here.

Reply:
I do not think this specific test code adds any value, as it is essentially testing standard library functionality (does printing an integer work?). Actual logic happens in the `From::from` implementation, and that's what should be tested.

I agree that my example of a generic implementation is very ugly due to the higher ranked trait bounds, however it can be greatly simplified using the `BitFields` trait idea I mentioned in another comment:

```rs
trait BitField { type Base; fn value(&self) -> &Self::Base; } // with `value` replacing the `From` impl
struct BitRange<'a, const START, const END, B: BitField>(&'a B::Base);

impl<'a, const START: u8, const END: u8, B: BitField> std::fmt::Display for BitRange<'a, B, START, END> where B::Base: std::fmt::Display {...}
```

Furthermore, if we move implementations from being macro generated to generic trait implementation, we improve extensibility (as it makes custom BitFields, for example based on `[u8; N]` or custom types, which cannot be generated by macros, easier to implement). 

While the standard library does use macro generated code for trait impls, but mostly in places were generics are not powerful enough (e.g. arrays prior to const generics, or implementation for numeric types which are not unified by any trait)

Reply:
I can see a future implementation doing this well. This is a large rework of the currently implemented functionality in the PR. I would lean towards this as a potential future improvement rather than a current issue that would block merging.

Reply:
We synced off-line to explore this a bit more and realized that using generic implementations will result in code more unreadable than I anticipated.